packages/vats/src/core/chain-behaviors.js

```diff
@@ -21,16 +22,77 @@ import {
 
 import * as Collect from '@agoric/run-protocol/src/collect.js';
 import { makeBridgeManager as makeBridgeManagerKit } from '../bridge.js';
+import * as BRIDGE_ID from '../bridge-ids.js';
 
-import { callProperties } from './utils.js';
+import { callProperties, extractPowers } from './utils.js';
 import { makeNameHubKit } from '../nameHub.js';
 
 export { installOnChain as installPegasusOnChain } from '@agoric/pegasus/src/install-on-chain.js';
 
 const { details: X } = assert;
 const { keys } = Object;
 
-const NUM_IBC_PORTS = 3;
+const NUM_IBC_PORTS_PER_CLIENT = 3;
+
+/**
+ * This registers the code triggered by `agd tx gov submit-proposal
+ * swingset-core-eval permit.json code.js`.  It is the "big hammer" governance
+ * that allows code.js access to all powers permitted by permit.json.
+ *
+ * @param {BootstrapPowers} allPowers
+ */
+export const bridgeCoreEval = async allPowers => {
+  // We need all of the powers to be available to the evaluator, but we only
+  // need the bridgeManager to install our handler.
+  const {
+    consume: { bridgeManager: bridgeManagerP },
+  } = allPowers;
+  const bridgeManager = await bridgeManagerP;
+  if (!bridgeManager) {
+    // Not running with a bridge.
+    return;
+  }
+
+  // Register a coreEval handler over the bridge.
+  const handler = Far('coreHandler', {
+    async fromBridge(_srcID, obj) {
+      switch (obj.type) {
+        case 'CORE_EVAL': {
+          /**
+           * Type defined by `agoric-sdk/golang/cosmos/proto/agoric/swingset/swingset.proto` CoreEval.
+           *
+           * @type {{ evals: { json_permits: string, js_code: string }[]}}
+           */
+          const { evals } = obj;
+          return Promise.all(
+            evals.map(({ json_permits: jsonPermit, js_code: code }) =>
+              // Run in a new turn to avoid crosstalk of the evaluations.
+              Promise.resolve().then(() => {
+                const permit = JSON.parse(jsonPermit);
+                const powers = extractPowers(permit, allPowers);
+
+                // Inspired by ../repl.js:
+                const globals = harden({
+                  ...farExports,
+                  assert,
+                  console,
+                  powers,
+                });
+
+                // Evaluate the code in the context of the globals.
+                return new Compartment(globals).evaluate(code);
```

Comment:
Let's have the evaluated code return a function and pass the powers to that function, so that the evaluated code can be organized according to POLA rather than having powerful globals available ambiently.